user/caribou/module/src/DSO9254AEventConverter.cc

```diff
@@ -0,0 +1,377 @@
+  std::cout << "Loaded parameters from configuration file." << std::endl
+	    << "  pedStartTime = " << pedStartTime << std::endl
+	    << "  pedEndTime   = " << pedEndTime   << std::endl
+	    << "  ampStartTime = " << ampStartTime << std::endl
+	    << "  ampEndTime   = " << ampEndTime   << std::endl
+	    << "  generateRoot = " << generateRoot << std::endl << std::endl;
+  
+
+  // Data container:
+  caribou::pearyRawData rawdata;
+
+  // internal containers
+  std::vector<std::vector<std::vector<double>>> wavess;
+  std::vector<std::vector<double>> peds;
+  std::vector<std::vector<double>> amps;
+  std::vector<double> time;
+  uint64_t timestamp;
+
+  // waveform parameters
+  std::vector<uint64_t> np;
+  std::vector<double> dx;
+  std::vector<double> x0;
+  std::vector<double> dy;
+  std::vector<double> y0;
+
+  
+  // generate rootfile to write waveforms as TH1D
+  TFile * histoFile = nullptr;
+  if( generateRoot ){
+    histoFile = new TFile( "waveforms.root", "UPDATE" ); // UPDATE?
+    if(!histoFile) {
+      LOG(ERROR) << "ERROR: " << histoFile->GetName() << " can not be opened";
+      return false;
+    }
+  }
+						       
+  
+  // Retrieve data from event
+  if(ev->NumBlocks() == 1) {
+
+    
+    // all four scope channels in one data block
+    auto datablock = ev->GetBlock(0);
+    LOG(DEBUG) << "DSO9254A frame with "<< datablock.size()<<" entries";
+
+    
+    // Calulate positions and length of data blocks:
+    // FIXME FIXME FIXME by Simon: this is prone to break since you are selecting bits from a 64bit
+    //                             word - but there is no guarantee in the endianness here and you
+    //                             might end up having the wrong one. 
+    // Finn: I am not sure if I actually solved this issue! Lets discuss.    
+    rawdata.resize( sizeof(datablock[0]) * datablock.size() / sizeof(uintptr_t) );
+    std::memcpy(&rawdata[0], &datablock[0], sizeof(datablock[0]) * datablock.size() );
+
+    
+    // needed per event
+    uint64_t block_words;
+    uint64_t pream_words;
+    uint64_t chann_words;
+    uint64_t block_posit = 0;
+    uint64_t sgmnt_count = 1;
+
+    
+    // loop 4 channels
+    for( int nch = 0; nch < 4; nch++ ){      
+
+      std::cout << "Reading channel " << nch << std::endl;
+
+      // Obtaining Data Stucture
+      block_words = rawdata[block_posit + 0];
+      pream_words = rawdata[block_posit + 1];
+      chann_words = rawdata[block_posit + 2 + pream_words];
+      std::cout << "  " << block_posit << " current position in block\n";
+      std::cout << "  " << block_words << " words per segment\n";
+      std::cout << "  " << pream_words << " words per preamble\n";
+      std::cout << "  " << chann_words << " words per channel data\n";
+      std::cout << "  " << rawdata.size() << " size of data rawdata\n";
+           
+
+      // read preamble:
+      std::string preamble = "";
+      for(int i=block_posit + 2;
```

Comment:
Do we really need to ad this byte-by-byte? `std::string` has great constructors which should also take a memory position and a length or something...

Reply:
i think this is related to the memcpy, which we want to discuss any way